File: Java/remove-duplicates-from-sorted-list-ii.java
//  Remove Duplicates from Sorted List II

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

// Time O(N), Space O(1)
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if (head == null)
            return null;
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode pre = dummy, cur = head;
        while (cur != null) {
            while (cur.next!=null && cur.val==cur.next.val)
                cur = cur.next;
            if (pre.next == cur)
                pre = cur;
            else
                pre.next = cur.next;
            cur = cur.next;
        }
        return dummy.next;
    }
}

File: Leetcode150/remove-duplicates-from-sorted-list-ii.cc
//  Remove Duplicates from Sorted List II

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if (!head || !head->next)   return head;
        int pre_num = head->val;
        ListNode *cur = head->next, *pre = head;
        while (cur) {
            if (cur->val == pre_num) {
                pre->next = cur->next;
                cur = pre->next;
            }
            else {
                pre_num = cur->val;
                pre = cur;
                cur = cur->next;
            }
        }
        return head;
    }
};
